### PR TITLE
Updating testing documentation and exposing some testing utils

### PR DESCRIPTION
**Pull request by LFDanLu**
Closes 

Somewhat relies on https://github.com/adobe/react-spectrum/pull/5192, test docs copy will change if we don't want to bring that PR in as well

The main changes to review are the updates to `testing.mdx` and the addition of and updates to `@react-aria/test-utils `, `@react-spectrum/test-utils`, `@react-spectrum/test-utils-internal` (aka `dev/test-utils`). I'll pull those changes out into another PR if it makes it easier to see the code changes.

The rest of the changes are to get rid of `triggerPress`, `fireEvent` and other utils in favor of userEvent interactions.

Also verified that the test util packages work locally via verdaccio and separate test suite in a CRA.

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 🧢 Your Project:

RSP

So I don't forget, changing the render wrapper to include Provider breaks some stuff: https://github.com/adobe/react-spectrum/pull/5508/files#r1411433715, revisit later

### CODE REVIEW

File: packages/@react-spectrum/test-utils/src/renderOverride.tsx
Comment (LFDanLu) [inline packages/@react-spectrum/test-utils/src/renderOverride.tsx:44]: redefining `rerender` so that the user can provide new `providerProps` seems to cause previously queried elements to be come stale (aka `let button = getByRole('button');` will point to the pre-rerender version of button). Going to move this work to a [different branch ](https://github.com/adobe/react-spectrum/tree/provider_render_wrapper) to work on separately, prioritizing other utils first

File: packages/dev/test-utils/package.json
Comment (LFDanLu) [inline packages/dev/test-utils/package.json:2]: renamed so `@react-spectrum/test-utils` is the publicly exposed one

File: packages/dev/test-utils/src/mockImplementation.ts
Comment (LFDanLu) [inline packages/dev/test-utils/src/mockImplementation.ts:18]: Open question, do we feel like we want to move any of these other utils (SSR stuff, StrictModeWrapper render wrapper, the stuff in this file) to the public package? I ran into difficulties with the docs build trying to keep the renderOverride re-export of testing library's stuff in the public package.

File: packages/@react-aria/test-utils/package.json
Comment (LFDanLu) [inline packages/@react-aria/test-utils/package.json:1]: will eventually house the aria pattern specific utils for carrying out specific interactions

File: packages/@react-spectrum/autocomplete/docs/SearchAutocomplete.mdx
Comment (LFDanLu) [inline packages/@react-spectrum/autocomplete/docs/SearchAutocomplete.mdx:1]: Added "Testing" sections to components that frequently caused issues for people testing our components (aka had overlays, rendered differently in mobile vs desktop, virtualized components, long press/drag via useMove). Open to feedback, pretty minimal for now

File: packages/@react-aria/aria-modal-polyfill/test/index.test.tsx
Comment (LFDanLu) [inline packages/@react-aria/aria-modal-polyfill/test/index.test.tsx:1]: In many of the tests, I got rid of triggerPress and typeText where ever possible in favor of userEvent. Also got rid of fireEvent where I could as well.

File: packages/@react-aria/aria-modal-polyfill/test/index.test.tsx
Comment (LFDanLu) [inline packages/@react-aria/aria-modal-polyfill/test/index.test.tsx:28]: In some of the tests that didn't have `userEvent` added, I opted to put the userEvent.setup inline with each test case as per https://kentcdodds.com/blog/avoid-nesting-when-youre-testing. Not sure which I prefer tbh...

File: packages/@react-aria/test-utils/src/events.ts
Comment (reidbarber) [inline packages/@react-aria/test-utils/src/events.ts:18]: Can we get a JSDoc comment here?

File: packages/dev/docs/pages/react-spectrum/testing.mdx
Comment (reidbarber) [inline packages/dev/docs/pages/react-spectrum/testing.mdx:221]: Nit: 
```suggestion
depending on the other time-sensitive behavior you are testing. Please see [Jest's timer docs](https://jestjs.io/docs/timer-mocks) or the equivalent docs of your test frameworks for more information on how to do so.
```

File: packages/dev/docs/pages/react-spectrum/testing.mdx
Comment (reidbarber) [inline packages/dev/docs/pages/react-spectrum/testing.mdx:267]: Just a reminder to handle this before merging

File: packages/dev/docs/pages/react-spectrum/testing.mdx
Comment (reidbarber) [inline packages/dev/docs/pages/react-spectrum/testing.mdx:344]:
```suggestion
If you are using React 16 or 17, you may run into an issue where the ids generated by the React Spectrum components are changing on every snapshot. To remedy this, simply wrap your component in a [SSRProvider](../react-aria/SSRProvider.html).
```

**Review by reidbarber — CHANGES_REQUESTED** [review]
Looks great! Just a few minor comments.

File: packages/@react-aria/tag/test/useTagGroup.test.js
Comment (snowystinger) [inline packages/@react-aria/tag/test/useTagGroup.test.js:107]: how did this pass earlier? i'm guessing it's called with all three because we don't actually remove the previous selected tags?

File: packages/dev/docs/pages/react-spectrum/testing.mdx
Comment (snowystinger) [inline packages/dev/docs/pages/react-spectrum/testing.mdx:193]: so much indent

File: packages/dev/test-utils/src/mockImplementation.ts
Comment (snowystinger) [reply packages/dev/test-utils/src/mockImplementation.ts:18]: lets start with what we have, we can add more later

**Review by snowystinger — APPROVED** [review]

File: packages/dev/docs/pages/react-spectrum/testing.mdx
Comment (LFDanLu) [reply packages/dev/docs/pages/react-spectrum/testing.mdx:267]: Yup, will add a comment below so I don't forget

File: packages/dev/docs/pages/react-spectrum/testing.mdx
Comment (LFDanLu) [reply packages/dev/docs/pages/react-spectrum/testing.mdx:193]: oh thats weird, I don't see that locally in my IDE :/, lemme try to fix

File: packages/@react-aria/dnd/test/useDroppableCollection.test.js
Comment (LFDanLu) [reply packages/@react-aria/dnd/test/useDroppableCollection.test.js:322]: yeahhhh, I was already tired of doing so many small tweaks that find+replace couldn't quite handle well haha. Happy to handle this later

File: packages/@react-aria/tag/test/useTagGroup.test.js
Comment (LFDanLu) [reply packages/@react-aria/tag/test/useTagGroup.test.js:107]: you know, I don't actually recall how this used to pass because as you've noted the test component doesn't actually remove any of the tags with its `onRemove`. Lemme poke around and see if I can find out

Comment (LFDanLu) [reply packages/@react-aria/tag/test/useTagGroup.test.js:107]: Ah, I remember now. So the old test was firing keydown directly on the "laundry" tag which doesn't actually cause it to be selected, hence why the backspace operation only triggers for the single item via https://github.com/adobe/react-spectrum/blob/46ae3314a2e2c1ab856c82c8d9dc941e8c70016e/packages/%40react-aria/tag/src/useTag.ts#L69-L70. I simply modified the test to fall more inline with the "should support selection" intent of this test case so that the "laundry" tag is actually selected before `Backspace` is pressed.

**Comment by LFDanLu** [comment]
Reminder for self or anyone looking to merge: the current copy relies on https://github.com/adobe/react-spectrum/pull/5192, so don't merge until then

**Review by snowystinger — APPROVED** [review]

**Review by ktabors — APPROVED** [review]

**Review by reidbarber — APPROVED** [review]

**Review by snowystinger — APPROVED** [review]

**Pull request merged**
